File: formula-cs/FormulaTest/FormulaPerformanceTest.cs
using Formula;

namespace FormulaTest;

public class FormulaPerformanceTest
{
    private const int Iterations = 1000;
    private const string FormulaText = "@alpha AND (@beta OR @delta) AND @sigma AND (@omega >= 5)";
    
    [Test]
    public void ParsePerformance()
    {
        var startTime = DateTime.UtcNow;
        for (var i = 0; i < Iterations; i++)
        {
            Formula.Formula.Parse(FormulaText);
        }
        var endTime = DateTime.UtcNow;
        var total = endTime - startTime;
        var average = total / Iterations;
        
        Console.WriteLine($"Parse Average: {average.TotalMilliseconds} ms");
    }
    
    [Test]
    public void ResolvePerformance()
    {
        var formula = Formula.Formula.Parse(FormulaText);
        var context = new DataContext(new Dictionary<string, IResolvable>
        {
            ["alpha"] = Resolvable.Just("true"),
            ["beta"] = Formula.Formula.Parse("!@delta"),
            ["delta"] = Formula.Formula.Parse("true AND @alpha"),
            ["sigma"] = Formula.Formula.Parse("@alpha AND @beta"),
            ["omega"] = Resolvable.Just("22"),
        });
        for (var j = 0; j < 20000; j++)
        {
            context.Set($"key_{j}", $"value_{j}");
        }
        var startTime = DateTime.UtcNow;
        for (var i = 0; i < Iterations; i++)
        {
            formula.Resolve(context);
        }
        var endTime = DateTime.UtcNow;
        var total = endTime - startTime;
        var average = total / Iterations;
        
        Console.WriteLine($"Resolve Average: {average.TotalMilliseconds} ms");
    }
    
    [Test]
    public void DeepResolvePerformance()
    {
        const int depth = 1000;
        var formula = Formula.Formula.Parse($"@step_{depth}");
        var context = new DataContext(new Dictionary<string, IResolvable>
        {
            ["step_1"] = Resolvable.Just(1)
        });
        for (var j = 2; j <= depth; j++)
        {
            context.Set($"step_{j}", Formula.Formula.Parse($"@step_{j-1} + 1"));
        }
        var startTime = DateTime.UtcNow;
        for (var i = 0; i < Iterations; i++)
        {
            formula.Resolve(context);
        }
        var endTime = DateTime.UtcNow;
        var total = endTime - startTime;
        var average = total / Iterations;
        
        Console.WriteLine($"Deep Resolve Average: {average.TotalMilliseconds} ms");
    }
}

File: formula-java/src/test/java/org/formula/parse/FormulaPerformanceTest.java
package org.formula.parse;

import static org.formula.parse.assertions.ResolvedValueAssertions.assertResolvedValue;

import org.formula.ResolvedValue;
import org.formula.context.MutableDataContext;
import org.junit.jupiter.api.Test;

public class FormulaPerformanceTest {
    private static final int ITERATIONS = 1000;
    private static final String FORMULA_TEXT = "@alpha AND (@beta OR @delta) AND @sigma AND (@omega >= 5)";

    @Test
    void parsePerformance() {
        var startTime = System.nanoTime();
        for (var i = 0; i < ITERATIONS; i++) {
            Formula.parse(FORMULA_TEXT);
        }
        var endTime = System.nanoTime();
        var total = endTime - startTime;

        System.out.printf("Parse Average: %.2f µs%n", total / ITERATIONS * 0.001);
    }

    @Test
    void resolvePerformance() {
        var formula = Formula.parse(FORMULA_TEXT);
        var context = MutableDataContext.create()
                .set("alpha", "true")
                .set("beta", Formula.parse("!@delta"))
                .set("delta", Formula.parse("true AND @alpha"))
                .set("sigma", Formula.parse("@alpha AND @beta"))
                .set("omega", "12");

        for (var j = 0; j < 20000; j++) {
            context.set("key_"+j, "value_"+j);
        }
        var startTime = System.nanoTime();
        for (var i = 0; i < ITERATIONS; i++) {
            formula.resolve(context);
        }
        var endTime = System.nanoTime();
        var total = endTime - startTime;

        System.out.printf("Resolve Average: %.2f µs%n", total / ITERATIONS * 0.001);
    }

    @Test
    void deepResolvePerformance() {
        final int depth = 1000;
        var formula = Formula.parse("@step_" + depth);
        var context = MutableDataContext.create()
                .set("step_1", 1);
        for (var j = 2; j <= depth; j++) {
            context = context.set("step_"+j, Formula.parse("@step_%d + 1".formatted(j - 1)));
        }
        var startTime = System.nanoTime();
        var result = ResolvedValue.none();
        for (var i = 0; i < ITERATIONS; i++) {
            result = formula.resolve(context);
        }
        var endTime = System.nanoTime();
        var total = endTime - startTime;

        assertResolvedValue(result).hasValue(depth);
        System.out.printf("Deep Resolve Average: %.2f µs%n", total / ITERATIONS * 0.001);
    }
}
